### DISCUSSION TITLE
Simple Moving Average (SMA)

### DISCUSSION THREAD

**Discussion by DaveSkender** [DaveSkender/Stock.Indicators#240] [Indicators]
Community discussion about [Simple Moving Average (SMA)](https://dotnet.stockindicators.dev/indicators/Sma/#content)

**Comment by ushahx** [comment]
Hi Dave, I have a question about the params for SMA. I see History and period only, how about if I want to get the same 20 SMA for 15 min period while I am providing 1min data in the history param? Would the library calculate the different variations like 15 min, 30 day, 1 day periods? If not can this be added as a feature?

**Reply by DaveSkender** [reply]
I’m not quite sure I understand.  SMA will work with any frequency of `history` data; however, it expects consistent time periods.  Can you elaborate a bit more?

**Reply by ushahx** [reply]
May be I need to get clarifications on how SMA works for a 1 min period vs 5 min vs 15 min. For e.g. if I see the SMA value at 8am pst for a 1 min SMA or 5 min SMA or 15 min SMA on think or swim each of the close values are different. Not sure why? I was under the impression the close value should be the same at a particular point in time. That is the reason I was thinking if I can pass only 1 min data for past hour can I get a 1 min or 5min or 15min SMA by passing a parameter. This will also help avoid making several API request to the data source for different data sets.

On another note, which APIs do you recommend to get data? Its really difficult to get any API in the market that will give a good chunk of stocks, they all limit to 100 or 200 stocks and API calls a min.

Uttam

**Reply by DaveSkender** [reply]
An example, SMA(20) with 1min bars will average the last 20 minutes.  SMA(20) with 15-minute bars will average the last 5 hours.  Not sure if that helps.  In both cases it takes the last 20 bars; only, each dataset has a different frequency.

**Reply by DaveSkender** [reply]
Here are [a few suggestions on stock APIs](https://daveskender.github.io/Stock.Indicators/docs/GUIDE.html#where-can-i-get-historical-quote-data).  Note, these are "free" services recommendations.  I'm sure there are significantly better API services out there if you want to pay.

**Reply by ushahx** [reply]
> An example, SMA(20) with 1min bars will average the last 20 minutes. SMA(20) with 15-minute bars will average the last 5 hours. Not sure if that helps. In both cases it takes the last 20 bars; only, each dataset has a different frequency.

Thanks for clarification Dave. In this case would it be possible to just pass 1 min data to the SMA function in your indicator and still get the different variations of SMA & Period combinations? I guess I am asking for a feature upgrade of an overload method taking period as a param as well.

Thanks for sharing the stock API link, I have been using Alpaca but saw some gaps in their 1min data bars. I also posted on their slack channel with screenshots.

Also thanks for creating this project it is really useful. Appreciate the efforts.

**Reply by DaveSkender** [reply]
You’re welcome.  I’ll probably not take that on as a feature, but if you were to get 1-minute bars and do the parsing in your wrapper code, you’d still be able to get what you’re seeking.

**Reply by DaveSkender** [reply]
If you have 1 minute quotes, you can use our [aggregation utility](https://dotnet.stockindicators.dev/utilities/#resize-quote-history) to convert it to larger aggregate candle period sizes.

```c#
// 10-period SMA for 15 minute quotes,
// starting with 1 minute quotes
var sma10 = oneMinQuotes
  .Aggregate(PeriodSize.FifteenMinutes)
  .GetSma(lookbackPeriods: 10);

```

**Comment by DharmenderTongar** [comment]
I think it considers Close to calculate SMA. How to get SMA on Quote Open?

**Reply by DaveSkender** [reply]
```c#
var results = quotes
  .Use(CandlePart.Open)
  .GetSma(lookbackPeriods: 10);
```

**Reply by DharmenderTongar** [reply]
Thanks, I got it!
